File: api_sample/typescript/cdb-get-token/test_cdb_get_token.ts
// Copyright 2018-present Network Optix, Inc. Licensed under MPL 2.0: www.mozilla.org/MPL/2.0/
/**
 * Offline tests for cdb_get_token.ts. No network, no account needed.
 *
 * Run from this folder:  node --test test_cdb_get_token.ts
 */

import assert from "node:assert/strict";
import test from "node:test";

import type { FetchImpl } from "../nx-types.ts";

import {
  buildTokenRequest,
  getToken,
  resolveConfig,
  loadEnvFile,
  parseArgs,
  AuthError,
  ApiError,
  type CliArgs,
} from "./cdb_get_token.ts";

/** A fetch double that also records the calls it received. */
interface RecordingFetch extends FetchImpl {
  calls: Array<{ url: string; body: Record<string, unknown> | null }>;
}

interface FakeFetchOptions {
  status?: number;
  json?: Record<string, unknown> | null;
  text?: string;
}

/**
 * A fake fetch. Records the URL + parsed JSON body it was called with, and
 * returns a canned Response-like object.
 */
function fakeFetch(
  { status = 200, json = null, text = "" }: FakeFetchOptions = {},
): RecordingFetch {
  const calls: Array<{ url: string; body: Record<string, unknown> | null }> =
    [];
  const impl = (async (url, options) => {
    const bodyStr =
      options && typeof options.body === "string" ? options.body : null;
    calls.push({
      url: String(url),
      body: bodyStr ? (JSON.parse(bodyStr) as Record<string, unknown>) : null,
    });
    return {
      status,
      ok: status < 400,
      async json() {
        if (json === null) {
          throw new Error("no json");
        }
        return json;
      },
      async text() {
        return text;
      },
    } as Response;
  }) as RecordingFetch;
  impl.calls = calls;
  return impl;
}

/** Build a CliArgs object with sensible defaults for resolveConfig tests. */
function cliArgs(overrides: Partial<CliArgs> = {}): CliArgs {
  return {
    host: null,
    user: null,
    password: null,
    mfaCode: null,
    cloudSystemId: null,
    envFile: ".env",
    insecure: false,
    tokenOnly: false,
    ...overrides,
  };
}

// ---------------------------------------------------------------------------
// buildTokenRequest(): the request body
// ---------------------------------------------------------------------------

test("body has the minimal fields", () => {
  const body = buildTokenRequest("me@x.com", "pw");
  assert.deepEqual(body, {
    grant_type: "password",
    response_type: "token",
    client_id: "3rdParty",
    username: "me@x.com",
    password: "pw",
  });
});

test("body adds mfaCode only when given", () => {
  assert.ok(!("mfaCode" in buildTokenRequest("u", "p")));
  assert.equal(buildTokenRequest("u", "p", "123456").mfaCode, "123456");
});

test("body adds scope only when site id given", () => {
  assert.ok(!("scope" in buildTokenRequest("u", "p")));
  const body = buildTokenRequest("u", "p", null, "sys-1");
  assert.equal(body.scope, "cloudSystemId=sys-1");
});

// ---------------------------------------------------------------------------
// getToken(): the call + response handling
// ---------------------------------------------------------------------------

test("getToken success returns the token and calls the right URL", async () => {
  const f = fakeFetch({ json: { access_token: "nxcdb-xyz", expires_in: 3600 } });
  const data = await getToken("https://nxvms.com", "me@x.com", "pw", {
    fetchImpl: f,
  });

  assert.equal(data.access_token, "nxcdb-xyz");
  assert.equal(f.calls[0]?.url, "https://nxvms.com/cdb/oauth2/token");
  assert.equal(f.calls[0]?.body?.username, "me@x.com");
});

test("getToken trims a trailing slash in the host", async () => {
  const f = fakeFetch({ json: { access_token: "t" } });
  await getToken("https://nxvms.com/", "u", "p", { fetchImpl: f });
  assert.equal(f.calls[0]?.url, "https://nxvms.com/cdb/oauth2/token");
});

test("getToken raises AuthError on bad credentials", async () => {
  const f = fakeFetch({ status: 401, text: "no" });
  await assert.rejects(
    () => getToken("https://nxvms.com", "u", "p", { fetchImpl: f }),
    AuthError,
  );
});

test("getToken raises ApiError when token missing", async () => {
  const f = fakeFetch({ json: { something: 1 } });
  await assert.rejects(
    () => getToken("https://nxvms.com", "u", "p", { fetchImpl: f }),
    ApiError,
  );
});

test("getToken raises ApiError on a non-auth HTTP error", async () => {
  const f = fakeFetch({ status: 500, text: "boom" });
  await assert.rejects(
    () => getToken("https://nxvms.com", "u", "p", { fetchImpl: f }),
    ApiError,
  );
});

// ---------------------------------------------------------------------------
// config + CLI parsing
// ---------------------------------------------------------------------------

test("CLI flag overrides env var and .env file", () => {
  const config = resolveConfig(
    cliArgs({ user: "cli-user" }),
    { NX_CLOUD_USER: "file-user" },
    { NX_CLOUD_USER: "env-user" },
  );
  assert.equal(config.user, "cli-user");
});

test("env var overrides .env file when no CLI flag", () => {
  const config = resolveConfig(
    cliArgs(),
    { NX_CLOUD_USER: "file-user" },
    { NX_CLOUD_USER: "env-user" },
  );
  assert.equal(config.user, "env-user");
});

test(".env file is used when nothing else is set", () => {
  const config = resolveConfig(cliArgs(), { NX_CLOUD_USER: "file-user" }, {});
  assert.equal(config.user, "file-user");
});

test("loadEnvFile returns {} for a missing file", () => {
  assert.deepEqual(loadEnvFile("/no/such/.env"), {});
});

test("parseArgs handles --flag value, --flag=value, and booleans", () => {
  const flags = parseArgs([
    "--host",
    "https://nxvms.com",
    "--user=me@x.com",
    "--insecure",
    "--token-only",
  ]);
  assert.equal(flags.host, "https://nxvms.com");
  assert.equal(flags.user, "me@x.com");
  assert.equal(flags.insecure, true);
  assert.equal(flags.tokenOnly, true);
});

test("parseArgs throws on an unknown argument", () => {
  assert.throws(() => parseArgs(["--nope"]), /Unknown argument/);
});

File: api_sample/python/cdb-get-token/test_cdb_get_token.py
# Copyright 2018-present Network Optix, Inc. Licensed under MPL 2.0: www.mozilla.org/MPL/2.0/
"""
Offline tests for cdb_get_token.py. No network, no account needed.

Run from this folder:  pytest -v
"""

import argparse

import pytest

import cdb_get_token as sample


class FakeResponse:
    def __init__(self, status_code=200, json_data=None, text=""):
        self.status_code = status_code
        self._json = json_data
        self.text = text

    @property
    def ok(self):
        return self.status_code < 400

    def json(self):
        if self._json is None:
            raise ValueError("no json")
        return self._json


class FakeSession:
    def __init__(self, post=None):
        self.verify = None
        self._post = post
        self.post_url = self.post_json = None

    def post(self, url, json=None, timeout=None):
        self.post_url, self.post_json = url, json
        return self._post


# ---------------------------------------------------------------------------
# build_token_request(): the request body
# ---------------------------------------------------------------------------

def test_body_minimal_fields():
    body = sample.build_token_request("me@x.com", "pw")
    assert body == {
        "grant_type": "password",
        "response_type": "token",
        "client_id": "3rdParty",
        "username": "me@x.com",
        "password": "pw",
    }


def test_body_adds_mfa_code_only_when_given():
    assert "mfaCode" not in sample.build_token_request("u", "p")
    assert sample.build_token_request("u", "p", mfa_code="123456")["mfaCode"] == "123456"


def test_body_adds_scope_only_when_site_id_given():
    assert "scope" not in sample.build_token_request("u", "p")
    body = sample.build_token_request("u", "p", cloud_site_id="sys-1")
    assert body["scope"] == "cloudSystemId=sys-1"


# ---------------------------------------------------------------------------
# get_token(): the call + response handling
# ---------------------------------------------------------------------------

def test_get_token_success():
    session = FakeSession(post=FakeResponse(200, {"access_token": "nxcdb-xyz",
                                                  "expires_in": 3600}))
    data = sample.get_token("https://nxvms.com", "me@x.com", "pw", session=session)

    assert data["access_token"] == "nxcdb-xyz"
    assert session.post_url == "https://nxvms.com/cdb/oauth2/token"
    assert session.post_json["username"] == "me@x.com"


def test_get_token_trims_trailing_slash_in_host():
    session = FakeSession(post=FakeResponse(200, {"access_token": "t"}))
    sample.get_token("https://nxvms.com/", "u", "p", session=session)
    assert session.post_url == "https://nxvms.com/cdb/oauth2/token"


def test_get_token_bad_credentials_raises():
    session = FakeSession(post=FakeResponse(401, text="no"))
    with pytest.raises(sample.AuthError):
        sample.get_token("https://nxvms.com", "u", "p", session=session)


def test_get_token_missing_token_raises():
    session = FakeSession(post=FakeResponse(200, {"something": 1}))
    with pytest.raises(sample.ApiError):
        sample.get_token("https://nxvms.com", "u", "p", session=session)


# ---------------------------------------------------------------------------
# config
# ---------------------------------------------------------------------------

def test_cli_overrides_env(monkeypatch):
    monkeypatch.setenv("NX_CLOUD_USER", "env-user")
    args = argparse.Namespace(host=None, user="cli-user", password=None,
                              mfa_code=None, cloud_site_id=None)
    config = sample.resolve_config(args, {"NX_CLOUD_USER": "file-user"})
    assert config["user"] == "cli-user"
